### PR TITLE
doc: update coding guidelines

### PR DESCRIPTION
**Pull request by shiqingg**
- update and add some rules in coding guidelines

Tracked-On: #861
Signed-off-by: Shiqing Gao <shiqing.gao@intel.com>

### CODE REVIEW

File: doc/developer-guides/coding_guidelines.rst [hunk 1]
```diff
@@ -1099,6 +1134,27 @@ Compliant example::
        showcase_test = showcase++;
 
 
+ST-19: Array indexing shall be valid
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]: Better might be:  ``Array indexing shall be in-bounds``

File: doc/developer-guides/coding_guidelines.rst [hunk 2]
```diff
@@ -1099,6 +1134,27 @@ Compliant example::
...
+====================================
+
+When indexing is performed on an array, the index value shall be smaller than
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]: A bit clearer would be:
```
An array index value shall be between zero (for the first element)
and the array size minus one (for the last element).  Out-of-bound 
array references are an undefined behavior and must be avoided.
```

File: doc/developer-guides/coding_guidelines.rst [hunk 3]
```diff
@@ -1696,6 +1752,40 @@ Compliant example::
        char chr = showcase[1];
 
 
+TY-20: The actual parameter type shall be the same with the formal parameter type
```
Comment (dbkinder) [inline doc/developer-guides/coding_guidelines.rst]: change ``same with`` to ``same as``

**Comment by shiqingg** [comment]
@dbkinder Thanks for your suggestion David! New version has been pushed.

**Pull request merged**